presto-spark-classloader-interface/src/main/java/com/facebook/presto/spark/classloader_interface/PrestoSparkTaskProcessor.java

```diff
@@ -60,12 +60,16 @@ public Iterator<Tuple2<MutablePartitionId, T>> process(
     {
         int partitionId = TaskContext.get().partitionId();
         int attemptNumber = TaskContext.get().attemptNumber();
-        return taskExecutorFactoryProvider.get().create(
+        long startTimeInMs = System.currentTimeMillis();
+        IPrestoSparkTaskExecutorFactory iPrestoSparkTaskExecutorFactory = taskExecutorFactoryProvider.get();
+        long taskBootstrapTimeInMillis = System.currentTimeMillis() - startTimeInMs;
+        return iPrestoSparkTaskExecutorFactory.create(
```

Comment:
I see that we are computing bootstrap time here and then we make it part of TaskInfo object which is then serialized, added to `taskInfoCollector` accumulator and then processed on the driver. In order to do this, you had to change TaskInfo object and add additional field and do the whole plumbing for vanilla Presto as well. 

I am wondering if we could register a new LongAccumulator (similar to `taskInfoCollector`) and use this explicitly for recording bootstrap time on task. Accumulators framework in Spark allows you to add stats on the task and aggregate them on the driver. So, inside `PrestoSparkQueryExecutionFactory`, you can create a new accumulator 
```
LongAccumulator bootstrapTimeCollector = new LongAccumulator();
bootstrapTimeCollector.register(sparkContext, Option.empty(), false);
```

Then pass this accumulator as `taskInfoCollector` is passed to this class, add `taskBootstrapTimeInMillis` in this accumulator and then just aggregate the total bootstrap time across all tasks on the driver. This way you don't have to change any existing core Presto classes like TaskInfo, TaskContext etc and this implementation will only be applicable for PrestoOnSpark. WDYT?

Reply:
I like the suggestion of recording this time separately, instead of affecting the existing Presto only classes. 

As for the exact implementation, I have a different suggestion. Instead of this 

> You can create a new accumulator (LongAccumulator)
Then pass this accumulator as taskInfoCollector is passed to this class, add taskBootstrapTimeInMillis in this accumulator and then just aggregate the total bootstrap time across all tasks on the driver

I'm thinking we can measure / implement at the executor level, given startup happens once per-executor. Main reason being, if think about how the metric would be used - we would want to see the relative time spent in startup v/s doing computation (task-execution) in a given executor, instead of a given task.
So something like below maybe

```
CollectionAccumulator<PrestoSparkExecutorStats> executorStatsCollector = new CollectionAccumulator<>();
executorStatsCollector.register(sparkContext, Option.empty(), false);
```

```
public class PrestoSparkExecutorStats
{
    private final int stageId;
    private final String executorId;
    private final long bootstrapTime;
    private final long totalTime;
}
```

WDYT @pgupta2  ?

Reply:
Makes sense to me.. Also, `PrestoSparkExecutorStats` can be extended in future to pass more executor level stats, if required.

Reply:
Exactly. Cool. I will update.